src/python/bcc/table.py

```diff
@@ -399,19 +399,91 @@ def clear(self):
+            return (ct.c_uint32(0), None, None)
+
+        if not count:  # means alloc maximum size
+            count = self.max_entries
+        elif count < 1 or count > self.max_entries:
+            raise ValueError("Wrong count")
+
+        if alloc_k:
+            keys = (self.Key * count)()
+        if alloc_v:
+            values = (self.Leaf * count)()
+
+        return (ct.c_uint32(count), keys, values)
+
+    def _sanity_check_keys_values(self, keys=None, values=None):
+        """Check if the given keys or values have the right type and size.
+
+        Args:
+            keys (ct.Array): keys array to check
+            values (ct.Array): values array to check
+        Returns:
+            ct.c_uint32 : the size of the array(s)
+        Raises:
+            ValueError: If length of arrays is < 1 or > self.max_entries, or
+            when both arrays length are different.
+            TypeError: If the keys and values are not an instance of ct.Array
+        """
+        arr_len = 0
+        for elem in [keys, values]:
+            if elem:
+                if not isinstance(elem, ct.Array):
+                    raise TypeError
+
+                arr_len = len(elem)
+                if arr_len < 1 or arr_len > self.max_entries:
+                    raise ValueError("Array's length is wrong")
+
+        if keys and values:
+            # check both length are equal
+            if len(keys) != len(values):
+                raise ValueError("keys array length != values array length")
+
+        return ct.c_uint32(arr_len)
+
     def items_lookup_batch(self):
-        # batch size is set to the maximum
-        batch_size = self.max_entries
-        out_batch = ct.c_uint32(0)
-        keys = (type(self.Key()) * batch_size)()
-        values = (type(self.Leaf()) * batch_size)()
-        count = ct.c_uint32(batch_size)
+        """Lookup the key-value pairs related to the keys given as parameters.
+
+        Args:
+            None
+        Yields:
+            tuple: The tuple of (key,value) for every entries that have
+            been looked up.
+        Raises:
+            Exception: If bpf syscall has failed or is invalid
+        Notes: lookup batch on a keys subset is not supported by the kernel.
+        """
+        ct_cnt, ct_keys, ct_values = self._alloc_keys_values(alloc_k=True,
+                                                             alloc_v=True)
+
         res = lib.bpf_lookup_batch(self.map_fd,
                                    None,
-                                   ct.byref(out_batch),
-                                   ct.byref(keys),
-                                   ct.byref(values),
-                                   ct.byref(count)
+                                   ct.byref(ct_cnt),
+                                   ct.byref(ct_keys),
+                                   ct.byref(ct_values),
+                                   ct.byref(ct_cnt)
                                    )
```

Comment:
Think it more. I think we may need a loop here. The following is the scenario. Say there are five buckets, max_entries = 5,
   1. lookup_batch find there are 3 elements in bucket 0, and copied three key/values.
   2. lookup_batch finished bucket 0 and not lock bucket 1 yet, in such an interval, some other threads/processes may delete all 3 elements in bucket 0 and add 3 new elements in bucket 1.
   3. lookup_batch lock bucket 1 and find 3 elements, not able to copy all of them and hence returns to user space with 3 elements.

So at the end, we got 3 elements, out_batch = 1 which indicates next time to start with bucket 1 internally.
The above case should be really rare but it may still happen if there are intensive mixed lookup/update/delete operations. But with max_entries as a buffer size, we are guaranteed to make further progress at least one bucket at a time and the loop will eventually terminate.

For checking return code, a successful syscall return codes should be:
  0 indicating successful get some values and not finished yet
  errno.ENOENT: done
Any other error code indicates an error, we don't need to single out errno.EINVAL here.

You can take a look at linux:tools/testing/selftests/bpf/map_tests/htab_map_batch_ops.c for more details.